routing/http/server/server.go

```diff
@@ -141,16 +157,100 @@ func (s *server) findProviders(w http.ResponseWriter, httpReq *http.Request) {
 		writeErr(w, "FindProviders", http.StatusBadRequest, fmt.Errorf("unable to parse CID: %w", err))
 		return
 	}
-	providers, err := s.svc.FindProviders(httpReq.Context(), cid)
+
+	var handlerFunc func(w http.ResponseWriter, provIter iter.Iter[types.ProviderResponse])
+
+	var supportsNDJSON bool
+	var supportsJSON bool
+	accepts := httpReq.Header.Values("Accept")
+	if len(accepts) == 0 {
+		handlerFunc = s.findProvidersJSON
+	} else {
+		for _, accept := range accepts {
+			mediaType, _, err := mime.ParseMediaType(accept)
+			if err != nil {
+				writeErr(w, "FindProviders", http.StatusBadRequest, fmt.Errorf("unable to parse Accept header: %w", err))
+				return
+			}
+
+			switch mediaType {
+			case "application/json":
+				supportsJSON = true
+			case "application/x-ndjson":
+				supportsNDJSON = true
+			}
+		}
+
+		if supportsNDJSON && !s.disableNDJSON {
+			handlerFunc = s.findProvidersNDJSON
+		} else if supportsJSON {
+			handlerFunc = s.findProvidersJSON
+		} else {
+			writeErr(w, "FindProviders", http.StatusBadRequest, errors.New("no supported content types"))
+			return
+		}
+	}
+
+	provIter, err := s.svc.FindProviders(httpReq.Context(), cid)
 	if err != nil {
 		writeErr(w, "FindProviders", http.StatusInternalServerError, fmt.Errorf("delegate error: %w", err))
 		return
 	}
-	response := types.ReadProvidersResponse{Providers: providers}
-	writeResult(w, "FindProviders", response)
+
+	handlerFunc(w, provIter)
+}
+
+func (s *server) findProvidersJSON(w http.ResponseWriter, provIter iter.Iter[types.ProviderResponse]) {
+	var (
+		providers []types.ProviderResponse
+		i         int
+	)
+
+	for {
+		v, ok, err := provIter.Next()
+		if !ok {
+			break
+		}
+		if err != nil {
+			writeErr(w, "FindProviders", http.StatusInternalServerError, fmt.Errorf("delegate error on result %d: %w", i, err))
+		}
+		providers = append(providers, v)
+		i++
+	}
+	response := jsontypes.ReadProvidersResponse{Providers: providers}
+	writeJSONResult(w, "FindProviders", response)
+}
+
+func (s *server) findProvidersNDJSON(w http.ResponseWriter, provIter iter.Iter[types.ProviderResponse]) {
+	w.Header().Set("Content-Type", "application/x-ndjson")
```

Comment:
the same as the previous comment: Shall we use here the previously defined constant?